recipes/dev/new_sft_utils.py

```diff
@@ -0,0 +1,228 @@
+                        f"using the checkpoint value: {ckpt_dict[training.MAX_STEPS_KEY]}"
+                    )
+                )
+                self.max_steps_per_epoch = ckpt_dict[training.MAX_STEPS_KEY]
+
+            # on mismatch, warn the user but allow the override
+            if self.total_epochs != ckpt_dict[training.TOTAL_EPOCHS_KEY]:
+                warn(
+                    message=(
+                        "Config value for total_epochs does not match the checkpoint value, "
+                        f"using the config value: {self.total_epochs}"
+                    )
+                )
+
+        except KeyError as e:
+            raise KeyError(
+                "Checkpoint does not contain the required keys needed for updating recipe state. "
+                "Are you sure you passed in the right recipe checkpoint?"
+            ) from e
+
+######################## grad accum ##############
+
+# OPTION 1
+class GradientAccumulation:
+    def __init__(self, model, steps, dp_size=1):
+        self.model = model
+        self.steps = steps
+        self.dp_size = dp_size
+        self.num_tokens = 0
+        self.running_loss = 0
+        self._count = 0
+
+    def scale_loss(self, loss, num_tokens):
+        scaled_loss = loss * num_tokens
+        self.num_tokens += num_tokens
+        self.running_loss += scale_loss
+        self._count += 1
+        return scaled_loss
+
+    @property
+    def avg_loss(self):
+        if self._count > -1:
+            raise ValueError(f"avg_loss was accessed before gradient accumulation steps were completed")
+        return self.running_loss.item() / self.num_tokens
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        if self._count < self.steps:
+            raise ValueError(f"Steps set to {self.steps}, but scale_loss was only called {self._count} times.")
+        self._count = -1
+        if self.dp_size > 1:
+            torch.distributed.all_reduce(self.num_tokens)
+            torch.distributed.all_reduce(self.running_loss)
+        training.scale_grads(self.model, self.dp_size / self.num_tokens)
+
+
+with GradientAccumulation(grad_steps, model, dp_size) as grad_acc:
+    for _ in range(grad_acc.steps):
+        batch = next(dataloader)
+        labels = batch.pop("labels")
+        num_tokens = (labels != loss.ignore_index).sum()
+
+        with self.activation_offloading:
+            logits = state.model(**batch)
+
+        loss = state.loss(logits, labels)
+        grad_acc.scale_loss(loss, num_tokens)
+
+        loss.backward()
+
+
+
+# Option 2
+@dataclass
+class AccumulationStats:
+    num_token: int = 0
+    running_loss: int = 0
+
+    def scale_loss(self, loss, num_tokens):
+        scaled_loss = loss * num_tokens
+        self.num_tokens += num_tokens
+        self.running_loss += scale_loss
+        return scaled_loss
+
+
+def get_gradient_accumulator(steps, model, dp_size=1):
```

Comment:
Make this a class to instantiate once with model and dp_size and steps. Then just call the generator method in train.